### PR TITLE
Draft of driver metrics

### PR DESCRIPTION
**Pull request by zhenlineo**
Added Basics of Driver metrics and Connection pool metrics.
The driver metrics are enabled with system property `driver.metrics.enabled=true`

### CODE REVIEW

File: driver/src/main/java/org/neo4j/driver/internal/DriverFactory.java [hunk 1]
```diff
@@ -77,18 +82,19 @@ public final Driver newInstance( URI uri, AuthToken authToken, RoutingSettings r
...
         RetryLogic retryLogic = createRetryLogic( retrySettings, eventExecutorGroup, config.logging() );
 
-        ConnectionPool connectionPool = createConnectionPool( authToken, securityPlan, bootstrap, config );
+        InternalAbstractDriverMetrics metrics = createDriverMetrics( config );
+        ConnectionPool connectionPool = createConnectionPool( authToken, securityPlan, bootstrap, metrics, config );
 
         InternalDriver driver = createDriver( uri, address, connectionPool, config, newRoutingSettings,
                 eventExecutorGroup, securityPlan, retryLogic );
 
+        driver.driverMetrics( metrics );
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/DriverFactory.java]: Would be nice to pass metrics into the driver constructor and avoid this setter.

File: driver/src/main/java/org/neo4j/driver/internal/InternalDriver.java [hunk 2]
```diff
@@ -144,6 +146,16 @@ private void assertOpen()
         }
     }
 
+    public DriverMetrics driverMetrics()
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/InternalDriver.java]: Maybe `#metrics()`?

File: driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import org.neo4j.driver.v1.Logging;
+
+import static org.neo4j.driver.internal.async.ChannelAttributes.serverAddress;
+
+public class NettyChannelTracker implements ChannelPoolHandler
+{
+    private final Map<BoltServerAddress,AtomicInteger> addressToInUseChannelCount = new ConcurrentHashMap<>();
+    private final Map<BoltServerAddress,AtomicInteger> addressToIdleChannelCount = new ConcurrentHashMap<>();
+    private final Logger log;
+    private DriverMetricsListener metricsListener;
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java]: Field can be final

File: driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    public void channelClosed( Channel channel )
+    {
+        decrementIdle( channel );
+        metricsListener.afterClosed( serverAddress( channel ) );
+    }
+
+    public int inUseChannelCount( BoltServerAddress address )
+    {
+        AtomicInteger count = addressToInUseChannelCount.get( address );
+        return count == null ? 0 : count.get();
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java:99]: Maybe `return addressToInUseChannelCount.getOrDefault(address, 0)`?

File: driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    public int idleChannelCount( BoltServerAddress address )
+    {
+        AtomicInteger count = addressToIdleChannelCount.get( address );
+        return count == null ? 0 : count.get();
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java:105]: Maybe `return addressToIdleChannelCount.getOrDefault(address, 0)`?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/HistogramSnapshot.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.metrics;
+
+import org.neo4j.driver.internal.metrics.spi.Histogram;
+
+public class HistogramSnapshot implements Histogram
+{
+    private Histogram copy;
+    private Histogram origin;
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/HistogramSnapshot.java]: Fields can be final

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalConnectionPoolMetrics.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import static org.neo4j.driver.internal.metrics.InternalDriverMetrics.serverAddressToUniqueName;
+
+public class InternalConnectionPoolMetrics implements ConnectionPoolMetrics, ConnectionPoolMetricsListener
+{
+    private final BoltServerAddress address;
+    private final ConnectionPool pool;
+
+    private AtomicLong created = new AtomicLong();
+    private AtomicLong closed = new AtomicLong();
+    private AtomicInteger toCreate = new AtomicInteger();
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/InternalConnectionPoolMetrics.java]: Why is this field `AtomicInteger` while others are `AtomicLong`?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalDriverMetrics.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        this.config = config;
+        this.connectionPoolMetrics = new ConcurrentHashMap<>();
+        this.connectionMetrics = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void addMetrics( BoltServerAddress serverAddress, ConnectionPoolImpl pool )
+    {
+        addPoolMetrics( serverAddress, pool );
+        addConnectionMetrics( serverAddress );
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/InternalDriverMetrics.java]: Can there be a race when pool metrics are added but connection metrics are not, and reader sees one but not the other?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    public String toString()
+    {
+        return format("[min=%sns, max=%sns, mean=%sns, stdDeviation=%s, totalCount=%s]",
+                min(), max(), mean(), stdDeviation(), totalCount());
+    }
+
+    public String printDistribution()
+    {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        PrintStream writer = new PrintStream( stream );
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java]: Can `StringWriter` be used instead?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Override
+    public void reset()
+    {
+        delegate.reset();
+    }
+
+    public static ConcurrentHistogram createHdrHistogram( long highestTrackableValue )
+    {
+        return new ConcurrentHistogram( highestTrackableValue, 3 );
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java]: Would be nice to have a constant for `3`

File: driver/src/main/java/org/neo4j/driver/internal/metrics/ListenerEvent.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.neo4j.driver.internal.metrics;
+
+
+public interface ListenerEvent
+{
+    void start();
+    long elapsed();
+
+    interface PoolListenerEvent extends ListenerEvent
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/ListenerEvent.java]: Are these two inner interfaces needed? Can we just keep `ListenerEvent`?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/spi/ConnectionPoolMetrics.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    int idle();
+
+    /**
+     * The amount of connections that is going to be created.
+     * The amount is increased by one when the pool noticed a request to create a new connection.
+     * The amount is decreased by one when the pool noticed a new connection is created regardless successfully or not.
+     * The number is changing up and down from time to time.
+     * @return The amount of connection that is to be created.
+     */
+    int toCreate();
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/spi/ConnectionPoolMetrics.java]: What do you think about `#creating()` name instead?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/spi/DriverMetrics.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.metrics.spi;
+
+import java.util.Map;
+
+public interface DriverMetrics
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/spi/DriverMetrics.java]: Maybe just `Metrics`?

File: driver/src/main/java/org/neo4j/driver/internal/metrics/spi/DriverMetrics.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+{
+    // TODO Once this interface become public, find a better way to enable metrics and detect metrics availability.
+    String DRIVER_METRICS_ENABLED_KEY = "driver.metrics.enabled";
+    static boolean isDriverMetricsEnabled()
+    {
+        return Boolean.valueOf( System.getProperty( DRIVER_METRICS_ENABLED_KEY, "false" ) );
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/spi/DriverMetrics.java]: `return Boolean.getBoolean(DRIVER_METRICS_ENABLED_KEY);` can be done instead

File: driver/src/main/java/org/neo4j/driver/internal/metrics/spi/ConnectionPoolMetrics.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+     * An unique name that identifies this connection pool metrics among all others
+     * @return An unique name
+     */
+    String uniqueName();
+
+    /**
+     * The status of the pool
+     * @return The status of the pool in a string
+     */
+    String poolStatus();
```
Comment (lutovich) [inline driver/src/main/java/org/neo4j/driver/internal/metrics/spi/ConnectionPoolMetrics.java]: I'd vote for an enum instead

**Review by lutovich — COMMENTED** [review]
Review round completed

File: driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java (hunk 4 above)
Comment (zhenlineo) [reply driver/src/main/java/org/neo4j/driver/internal/async/pool/NettyChannelTracker.java:99]: I cannot default to 0 as it is an `Integer` object. I feel it is more expensive to do extra boxing and unboxing than this null check.

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalConnectionPoolMetrics.java (hunk 7 above)
Comment (zhenlineo) [reply driver/src/main/java/org/neo4j/driver/internal/metrics/InternalConnectionPoolMetrics.java]: The diff between `Gauges` (int) and `Counters` (long)

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java (hunk 9 above)
Comment (zhenlineo) [reply driver/src/main/java/org/neo4j/driver/internal/metrics/InternalHistogram.java]: Removed this not used method.

File: driver/src/main/java/org/neo4j/driver/internal/metrics/InternalDriverMetrics.java (hunk 8 above)
Comment (zhenlineo) [reply driver/src/main/java/org/neo4j/driver/internal/metrics/InternalDriverMetrics.java]: Internally we will not access any of these metrics until this method returns and a connection pool is created. 

Externally, a user could query the metrics between the time slots where pool metrics are added, but not yet connection metrics. But I guess that does not matter too much? As it is a snapshot-ish thing to external users anyway. If they happens read at that time, then they just get half of the whole view. Then next time when they query the metrics, they will see the newly created ones.

**Pull request merged**